store/postgres/src/store_events.rs

```diff
@@ -0,0 +1,141 @@
+    }
+}
+
+impl EventProducer<StoreEvent> for StoreEventListener {
+    fn take_event_stream(&mut self) -> Option<StoreEventStreamBox> {
+        self.notification_listener.take_event_stream().map(
+            |stream| -> Box<Stream<Item = _, Error = _> + Send> {
+                Box::new(stream.map(|notification| {
+                    // Parse notification as JSON
+                    let value: serde_json::Value = serde_json::from_str(&notification.payload)
+                        .expect(&format!(
+                            "invalid JSON store event received from database: '{}'",
+                            &notification.payload
+                        ));
+
+                    // Create StoreEvent from JSON
+                    let change: StoreEvent =
+                        serde_json::from_value(value.clone()).unwrap_or_else(|_| {
+                            panic!("invalid store event received from database: {:?}", value)
+                        });
+
+                    change
+                }))
+            },
+        )
+    }
+}
+
+// Querying the entity changes corresponding to the reversion of a certain
+// block for a certain subgraph. We want to keep the deserialization
+// logic (QueryableByName) close to the actual query. That's why we introduce
+// the intermediate EntityChangeQBN struct
+
+struct EntityChangeQBN(EntityChange);
+
+impl QueryableByName<Pg> for EntityChangeQBN {
+    fn build<R: diesel::row::NamedRow<diesel::pg::Pg>>(
+        row: &R,
+    ) -> diesel::deserialize::Result<Self> {
+        let subgraph_id: String = row.get("subgraph")?;
+        let subgraph_id = SubgraphDeploymentId::new(subgraph_id).map_err(|_| {
+            format!(
+                "Illegal subgraph_id {}",
+                row.get("subgraph").unwrap_or("<unknown>".to_string())
+            )
+        })?;
+        let change: String = row.get("change")?;
+        Ok(EntityChangeQBN(EntityChange {
+            subgraph_id,
+            entity_type: row.get("entity")?,
+            entity_id: row.get("entity_id")?,
+            operation: match change.as_str() {
+                "set" => EntityChangeOperation::Set,
+                "removed" => EntityChangeOperation::Removed,
+                _ => return Err(Box::new(format_err!("bad operation {}", change).compat())),
+            },
+        }))
+    }
+}
+
+// Get the store event corresponding to reverting the operations for
+// block block_ptr_from. After reversion, we will be at block_ptr_to, which
+// must be one less than block_ptr_from. The source of the event will be
+// block_ptr_to as we will be at that block after the reversion
+pub fn get_revert_event(
+    conn: &PgConnection,
+    subgraph_id: &SubgraphDeploymentId,
+    block_ptr_from: &EthereumBlockPointer,
+    block_ptr_to: EthereumBlockPointer,
+) -> Result<StoreEvent, StoreError> {
+    // Sanity check on block numbers
+    if block_ptr_from.number != block_ptr_to.number + 1 {
```

Comment:
I'd use `assert_eq!` here. Or, if we don't care about this in production builds (I think we _do_), `debug_assert_eq!`.